### ISSUE TITLE
Firefox Version

### ISSUE THREAD

**Issue by RobotsLikeKetchup** [chinese-words-separator/chinese-words-separator.github.io#30]
I would like to use this extension on Firefox, as it is very useful.

I tried to port it for Firefox, just by changing the manifest for browser-specific things and uploading it as a temporary add-on, but there are errors that I have no idea how to fix.

If possible, could you create a version of this that works for firefox?

**Comment by chinese-words-separator** [comment]
Functionalities that are not working on Firefox but not limited to:

* Text-to-speech on mouse hover
    Cause: chrome.tts API is not supported on Firefox. Chrome and Edge does

* Parallel Text
    Cause: Firefox has no built-in translator. Chrome, Safari, and Edge does

* Respeak
    Cause: SpeechRecognition API is not supported on Firefox. Chrome and Safari does

* Mark correctly spoken
    Cause: SpeechRecognition API is not supported on Firefox. Chrome and Safari does

There is a worry that if Chinese word separator is ported to Firefox, and yet many functionalities are not working, some users might feel dissatisfied that Firefox is not well-supported and they might mistakenly post their comments and ratings to Chrome web store

Unless the browser APIs mentioned above is supported by Firefox, the port of CWS on Firefox will feel subpar

For the meantime, use CWS on Chrome, Safari, and Edge

And also, the marketshare of Firefox is abysmal. Perhaps some of it can be attributed to Firefox not having support on some browser APIs

![image](https://github.com/user-attachments/assets/8a08014a-3043-4d60-b881-d8672293cc73)

Whatever success Chrome is reaping now, we can attribute some of it to them gaining the best developers on the planet.

"To make that happen, [Google "poached" top web browser developers from Firefox, such as Ian Hickson, Darin Fisher, Pam Greene, and Brian Ryner.](https://www.niallkennedy.com/blog/2008/09/google-chrome.html) Both Microsoft and the Mozilla Foundation were caught flat-footed. Neither have caught up." -- https://www.zdnet.com/home-and-office/networking/the-fall-of-firefox-mozillas-once-popular-web-browser-slides-into-irrelevance/

Anyway, Chinese words separator will be ported to Firefox regardless of Firefox's market share, when all the browser APIs mentioned above are supported by Firefox

By the way if you need to use CWS on your mobile, you can use CWS by using browsers that supports extensions, e.g., [Kiwi](https://play.google.com/store/apps/details?id=com.kiwibrowser.browser&hl=en), [Lemur](https://play.google.com/store/apps/details?id=com.lemurbrowser.exts&hl=en), then install CWS from [Chrome web store](https://chrome.google.com/webstore/detail/chinese-words-separator/gacfacdpfimbkgcnlegknnmcccjgcbnp). Some users think that extensions on Android are supported on Firefox only, as Chrome on Android does not support extensions. You can also use CWS on Safari of [iPhone/iPad/Mac](https://apps.apple.com/us/app/chinese-words-separator/id1598790017)

Don't upload CWS to Firefox add-ons store, just use it specifically on your own machine, don't distribute it, due to the worry I mentioned above. Besides, CWS's copyright belongs to me

**Comment by RobotsLikeKetchup** [comment]
Yeah I wasn't going to upload it or publish it, I just wanted to use it on my own computer.

Its just that my computer for some reason runs firefox way better than chrome- chrome makes my computer run out of battery very very fast.

I don't need the text-to-speech features right now anyways, but when I tried to just copy the files and put them in firefox(and changed the manifest.js file) I got errors that are way out of my knowledge to fix-all of them were that the runtime.sendMessage didn't work because it was in the window frame instead of the globalServiceWorker frame.

Is there a way to fix this? Otherwise, I will just use this extension on microsoft edge.

Thank you so much for making this extension free by the way. It is obvious that you made this app with a lot of time and effort and it is really a useful app, probably one of the coolest ways to help learn chinese.

**Comment by shamefulCake1** [comment]
Firefox seems to have a built-in translator now.

Anyway, Firefox is the only browser supporting extensions on Android, and also the only one which supports Manifest V2, which means true adblocking, and a lot of superior features.

**Comment by chinese-words-separator** [comment]
Thanks for the update on Firefox. There are other APIs that Firefox still doesn't support, e.g., SpeechRecognition

There are browsers in Android that supports extensions, e.g., Lemur and Microsoft Edge Canary. I put the instruction on Chrome Web Store. I'll put it here too:

Note: Kiwi browser is discontinued since January 2025, some of Kiwi's code is now integrated to Microsoft Edge Canary Android browser. Follow the instruction here to install extensions on Microsoft Edge Canary Android browser: https://github.com/kiwibrowser/src.next. CWS's extension id on Microsoft Edge Add-ons store is peglphdebebicogkcpinemlaanfabeip

If installing extensions on Microsoft Edge Canary Android browser is cumbersome, you can opt to use Lemur browser instead on your Android device:
https://play.google.com/store/apps/details?id=com.lemurbrowser.exts

then install CWS from Chrome Web Store.

To use the extension on Safari on iOS, iPadOS, and macOS, install CWS from:
https://apps.apple.com/app/chinese-words-separator/id1598790017
